compiler/rustc_const_eval/src/interpret/discriminant.rs

```diff
@@ -286,13 +314,25 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 ..
             } => {
                 assert!(variant_index != untagged_variant);
+                let discr_len = (niche_variants.end().as_u32() - niche_variants.start().as_u32())
+                    .checked_add(1)
+                    .expect("the number of niche variants fits into u32");
                 let variants_start = niche_variants.start().as_u32();
-                let variant_index_relative = variant_index
+                let adj_idx = variant_index
                     .as_u32()
                     .checked_sub(variants_start)
                     .expect("overflow computing relative variant idx");
+
+                let variant_index_relative = if niche_variants.contains(&untagged_variant) {
+                    let adj_untagged_idx = untagged_variant.as_u32() - variants_start;
+                    let adj_idx_to_modulo = adj_idx
+                        .checked_add(discr_len - adj_untagged_idx)
+                        .expect("overflow computing relative variant idx");
+                    adj_idx_to_modulo % discr_len - 1
+                } else {
+                    adj_idx
+                };
                 // We need to use machine arithmetic when taking into account `niche_start`:
-                // tag_val = variant_index_relative + niche_start_val
```

Comment:
Why did you remove this comment? The code below still does the same math as before, doesn't it?

Reply:
I probably didn't notice that this is a comment rather than outdated code to be deleted.